File: protocol-generator/swift/Tests/EntityCoreProtocolTests/ScopeAlgebraTests.swift
// ScopeAlgebraTests.swift — the 0.8.2.20/.21/.24/.25 surface: §5.2 effective
// targets, §6.3 `check_path_permission`, §5.4's path-scope-only sentinel, F50's
// typing of `scope_subset`, and §4.11/§5.2a's code-belongs-to-the-cause split.
//
// The oracle carries no vector on most of this — the 778-check set is indifferent
// to every function below — so the accept direction is the peer's own to cover.
// Each mutation control that must redden a named case here lives in
// output/scratch/plant-swift.sh and is EXECUTED, not described.

import XCTest
@testable import EntityCoreProtocol

final class ScopeAlgebraTests: XCTestCase {

    // A syntactically valid peer_id (>= 46 Base58 chars) so `canonicalize` and
    // `isPeerID` behave as they do on the wire.
    let local = "12D3KooWLocalPeerIdExampleAAAAAAAAAAAAAAAAAAAA"

    func scope(_ include: [String], exclude: [String] = []) -> Capability.Scope {
        Capability.Scope(include: include, exclude: exclude)
    }

    func grant(handlers: [String], operations: [String],
               resources: [String], resourcesExclude: [String] = []) -> CBORValue {
        func sc(_ incl: [String], _ excl: [String]) -> CBORValue {
            var f: [(String, CBORValue)] = [("include", .array(incl.map { .text($0) }))]
            if !excl.isEmpty { f.append(("exclude", .array(excl.map { .text($0) }))) }
            return .textMap(f)
        }
        return .textMap([
            ("handlers", sc(handlers, [])),
            ("operations", sc(operations, [])),
            ("resources", sc(resources, resourcesExclude)),
        ])
    }

    func token(_ grants: [CBORValue]) -> Entity {
        Entity(type: "system/capability/token", data: .textMap([("grants", .array(grants))]))
    }

    func execute(targets: [String]? = nil, exclude: [String]? = nil,
                 targetsRaw: CBORValue? = nil, resourceOmitsTargets: Bool = false) -> Entity {
        var fields: [(String, CBORValue)] = [("operation", .text("get"))]
        if resourceOmitsTargets {
            fields.append(("resource", .textMap([("exclude", .array([]))])))
        } else if let targetsRaw {
            fields.append(("resource", .textMap([("targets", targetsRaw)])))
        } else if let targets {
            var r: [(String, CBORValue)] = [("targets", .array(targets.map { .text($0) }))]
            if let exclude { r.append(("exclude", .array(exclude.map { .text($0) }))) }
            fields.append(("resource", .textMap(r)))
        }
        return Entity(type: "system/protocol/execute", data: .textMap(fields))
    }

    // MARK: §5.2 effective targets (0.8.2.20/.21, N11)

    /// The caller's own exclude removes entries BEFORE anything else looks at the
    /// request, and survivors keep the caller's RAW spelling (0.8.2.21) rather than a
    /// canonical form.
    func testCallerExcludeRemovesItsOwnTarget() {
        XCTAssertEqual(Capability.effectiveTargets(execute(targets: ["qA", "qB"], exclude: ["qB"]),
                                                   localPeerID: local) ?? [], ["qA"])
    }

    func testSurvivorsAreRawNotCanonicalized() {
        XCTAssertEqual(Capability.effectiveTargets(execute(targets: ["qA"]), localPeerID: local) ?? [], ["qA"])
    }

    /// N11's NON-LOSSY PROJECTION `[MUST]`: the two empties are DIFFERENT REQUESTS and a
    /// function returning only a list cannot tell them apart. A resource-OPTIONAL
    /// operation answers them differently (N7/N10 + EXTENSION-TREE §2.2a: absent → root
    /// listing, self-excluded → 400 path_required), so collapsing them here would delete
    /// the discriminator before any handler could read it.
    func testAbsentResourceIsNil() {
        XCTAssertNil(Capability.effectiveTargets(execute(), localPeerID: local))
    }

    func testSelfExcludedResourceIsEmptyButPresent() {
        let eff = Capability.effectiveTargets(execute(targets: ["qA"], exclude: ["qA"]), localPeerID: local)
        XCTAssertNotNil(eff, "a PRESENT resource whose every target was excluded must not read as absent")
        XCTAssertEqual(eff ?? ["unset"], [])
    }

    /// PRESENT-BUT-ILL-TYPED `targets` IS PRESENT — the cell the two vanguards disagreed
    /// on until 0.8.2.25, corrected toward `go`. Reading it as ABSENT would serve a
    /// present resource the wider absent-case answer §3.3 forbids (on `get`, the root
    /// listing instead of 400 path_required). The KEY's presence is the discriminator,
    /// not the value's type.
    func testIllTypedTargetsIsPresentWithAnEmptyEffectiveList() {
        let eff = Capability.effectiveTargets(execute(targetsRaw: .text("qA")), localPeerID: local)
        XCTAssertNotNil(eff, "an ill-typed `targets` is PRESENT, not absent")
        XCTAssertEqual(eff ?? ["unset"], [])
    }

    func testResourceWithNoTargetsKeyIsAbsent() {
        XCTAssertNil(Capability.effectiveTargets(execute(resourceOmitsTargets: true), localPeerID: local))
    }

    /// §5.4's caller-exclude arm is fail-OPEN on an unmatchable pattern (the GRANT arm is
    /// fail-CLOSED): `canonicalize` answers the sentinel, `matchesPattern` then answers
    /// false, and the target simply SURVIVES.
    func testUnmatchableCallerExcludeCarvesOutNothing() {
        XCTAssertEqual(Capability.effectiveTargets(execute(targets: ["qA"], exclude: ["../nope"]),
                                                   localPeerID: local) ?? [], ["qA"])
    }

    // MARK: §6.3 check_path_permission

    func cpp(_ tok: Entity, _ path: String, op: String = "get", pattern: String = "system/tree") -> Bool {
        Capability.checkPathPermission(operation: op, path: path, handlerPattern: pattern,
                                       grants: Capability.grants(of: tok), localPeerID: local)
    }

    var okToken: Entity {
        token([grant(handlers: ["system/tree"], operations: ["get"], resources: ["q/*"])])
    }

    /// THE ACCEPT CASE IS WHAT VALIDATES THE FIXTURE. A predicate test built only from
    /// deny cases is indistinguishable from one asserting `false == false`, and a broken
    /// fixture makes every deny pass for free.
    func testAcceptValidatesTheFixture() {
        XCTAssertTrue(cpp(okToken, "q/a"))
    }

    /// One deny per DIMENSION: a single deny cannot separate "it checks the dimension I
    /// care about" from "it denies".
    func testDenyOnResourcesDimension() { XCTAssertFalse(cpp(okToken, "other/a")) }
    func testDenyOnOperationsDimension() { XCTAssertFalse(cpp(okToken, "q/a", op: "put")) }
    func testDenyOnHandlersDimension() { XCTAssertFalse(cpp(okToken, "q/a", pattern: "system/other")) }

    /// §5.2's note: an empty `resources.include` is a legal grant shape (a handler that
    /// touches no tree paths) and DENIES every path here.
    func testEmptyResourcesIncludeDeniesEveryPath() {
        XCTAssertFalse(cpp(token([grant(handlers: ["system/tree"], operations: ["get"], resources: [])]), "q/a"))
    }

    /// A malformed path canonicalizes to the sentinel, which matches no grant (§5.4), so
    /// it falls through to DENY rather than being matched at all.
    func testMalformedPathDenies() { XCTAssertFalse(cpp(okToken, "../nope")) }

    /// A grant exclude covering the subject denies: the caller's exclusions were already
    /// applied in deriving this concrete path, so nothing carves it back.
    func testGrantExcludeCoveringThePathDenies() {
        XCTAssertFalse(cpp(token([grant(handlers: ["system/tree"], operations: ["get"],
                                        resources: ["q/*"], resourcesExclude: ["q/a"])]), "q/a"))
    }

    // MARK: §5.4's sentinel is PATH-SCOPE only (0.8.2.24, N2/N3)

    /// The discriminating pair, and the id case cannot be passed by accident: `*/apply`
    /// is an ordinary namespaced operation name that PATH-canonicalizes to the sentinel,
    /// so on the pre-.24 unscoped reading it denied the WHOLE dimension and `get`
    /// included by a bare `*` came back false.
    func testSentinelDoesNotReachIDScopeOperations() {
        XCTAssertTrue(Capability.matchesScope("get", scope(["*"], exclude: ["*/apply"]),
                                              frame: local, kind: .id))
    }

    func testSentinelDoesNotReachIDScopePeers() {
        XCTAssertTrue(Capability.matchesScope(local, scope(["*"], exclude: ["../nope"]),
                                              frame: local, kind: .id))
    }

    /// The other half, and it proves this is a scope SPLIT rather than a removal: on a
    /// path-scope dimension an unmatchable exclude still denies (0.8.2.21), because there
    /// it would otherwise carve out nothing and leave the grant silently wider than its
    /// author wrote.
    func testSentinelStillDeniesOnPathScope() {
        XCTAssertFalse(Capability.matchesScope("system/tree", scope(["*"], exclude: ["../nope"]),
                                               frame: local, kind: .path))
    }

    func testOrdinaryPathScopeExcludeCarvesOutOnlyItsTarget() {
        XCTAssertTrue(Capability.matchesScope("system/tree", scope(["*"], exclude: ["system/secret"]),
                                              frame: local, kind: .path))
    }

    // MARK: F50 — scope_subset is typed by scope kind (ruled 0.8.2.16)

    func ssub(_ child: Capability.Scope, _ parent: Capability.Scope, _ kind: Capability.ScopeKind) -> Bool {
        Capability.scopeSubset(child, parent, childFrame: local, parentFrame: local, kind: kind)
    }

    /// The two pairs `entity-core-formalization` measured as disagreeing on `lean`, both
    /// fail-CLOSED under the canonicalizing reading. On the id matcher a literal child
    /// include is covered by a bare `*` parent.
    func testScopeSubsetIDCoversPathShapedLiteral() {
        XCTAssertTrue(ssub(scope(["/tree/get"]), scope(["*"]), .id))
    }

    func testScopeSubsetIDCoversStarSlashApply() {
        XCTAssertTrue(ssub(scope(["*/apply"]), scope(["*"]), .id))
    }

    /// The id matcher must still REFUSE a genuine widening, or "typed" would just mean
    /// "always true".
    func testScopeSubsetIDRefusesWidening() {
        XCTAssertFalse(ssub(scope(["*"]), scope(["get"]), .id))
    }

    /// And the path arm is UNCHANGED — this is a split, not a replacement.
    func testScopeSubsetPathStillCovers() {
        XCTAssertTrue(ssub(scope(["q/a"]), scope(["*"]), .path))
    }

    func testScopeSubsetPathRefusesUncovered() {
        XCTAssertFalse(ssub(scope(["q/a"]), scope(["r/*"]), .path))
    }

    // MARK: §5.2a / §4.11 — the code belongs to the cause (0.8.2.24 N4/N5, 0.8.2.25)

    /// A mis-keyed `included` entry is a RESOLUTION-INTEGRITY failure, not a CBOR
    /// tag-policy violation — its encoding is canonical, and what is false is the claim
    /// the KEY makes — so it MUST answer 400 `hash_mismatch` and `non_canonical_ecf` is
    /// explicitly non-conformant there.
    func testMisKeyedIncludedThrowsHashMismatch() throws {
        let inner = try Model.make(type: "primitive/any", fields: [("v", .text("x"))])
        let envelope = CBORValue.textMap([
            ("root", try CBOR.decode(Model.make(type: Wire.executeType, data: .map([])).bytes)),
            ("included", .map([(key: .bytes([UInt8](repeating: 0, count: 33)),
                                value: try CBOR.decode(inner.bytes))])),
        ])
        let bytes = try CBOR.encode(envelope)
        XCTAssertThrowsError(try Wire.decodeEnvelope(bytes)) { err in
            // Named rather than left to `XCTAssertThrowsError`'s bare "something threw":
            // the pre-0.8.2.24 spelling was `.malformed`, which would still satisfy an
            // unqualified assertion.
            guard case .hashMismatch = (err as? CodecError) ?? .badSeed else {
                return XCTFail("expected .hashMismatch, got \(err)")
            }
        }
    }

    func testPreAdmissionMisKeyedIsHashMismatch() {
        let r = Wire.preAdmissionRefusal(.hashMismatch("x"))
        XCTAssertEqual(r.status, 400); XCTAssertEqual(r.code, "hash_mismatch")
    }

    func testPreAdmissionTagKeepsNonCanonicalECF() {
        let r = Wire.preAdmissionRefusal(.tagRejected)
        XCTAssertEqual(r.status, 400); XCTAssertEqual(r.code, "non_canonical_ecf")
    }

    func testPreAdmissionStructuralIsInvalidRequest() {
        let r = Wire.preAdmissionRefusal(.malformed("x"))
        XCTAssertEqual(r.status, 400); XCTAssertEqual(r.code, "invalid_request")
    }

    func testPreAdmissionOversizeIs413() {
        let r = Wire.preAdmissionRefusal(.frameTooLarge)
        XCTAssertEqual(r.status, 413); XCTAssertEqual(r.code, "payload_too_large")
    }

    /// §4.11's framing arm is owed a coded frame; an ordinary close is NOT a refusal of
    /// anything and there is nobody left to answer.
    func testFramingRefusalClassification() {
        XCTAssertTrue(Wire.isFramingRefusal(.frameTooLarge))
        XCTAssertTrue(Wire.isFramingRefusal(.truncatedFrame))
        XCTAssertFalse(Wire.isFramingRefusal(.malformed("x")))
    }

    // MARK: the classification has to happen AT THE READ

    /// The three checks above are pure functions of an error VALUE: they stay green while
    /// `readFrame` reports the WRONG one. §4.11's distinction — "a clean close at a frame
    /// BOUNDARY is an ordinary hangup and is owed nothing; a stream that ends mid-frame
    /// is a REFUSAL and is owed a coded frame" — can only be made where the frame
    /// boundary is known, so it is driven here over a real socketpair. A test that never
    /// executes the site it is about is the never-executed-guard class wearing a green
    /// tick.
    func readAfter(_ sent: [UInt8]) -> Socket.FrameRead {
        var fds: [Int32] = [0, 0]
        XCTAssertEqual(socketpair(AF_UNIX, Int32(SOCK_STREAM.rawValue), 0, &fds), 0)
        let a = Socket(fd: fds[0]), b = Socket(fd: fds[1])
        if !sent.isEmpty { _ = sent.withUnsafeBytes { write(fds[1], $0.baseAddress!, sent.count) } }
        shutdown(fds[1], Int32(SHUT_WR))
        let r = a.readFrame()
        a.close(); b.close()
        return r
    }

    func testCleanCloseAtABoundaryIsNotARefusal() {
        guard case .closed = readAfter([]) else { return XCTFail("expected .closed") }
    }

    func testPartialLengthPrefixIsTruncated() {
        guard case .refused(.truncatedFrame) = readAfter([0, 0]) else {
            return XCTFail("expected .refused(.truncatedFrame)")
        }
    }

    func testPrefixDeclaringMoreThanArrivesIsTruncated() {
        guard case .refused(.truncatedFrame) = readAfter([0, 0, 0, 0x64, 0x61, 0x62]) else {
            return XCTFail("expected .refused(.truncatedFrame)")
        }
    }

    /// The body arm is the one an "off == 0" inference gets wrong: the prefix has been
    /// consumed, so ZERO body bytes is still a truncation and not a boundary.
    func testPrefixWithNoBodyAtAllIsStillTruncated() {
        guard case .refused(.truncatedFrame) = readAfter([0, 0, 0, 0x64]) else {
            return XCTFail("expected .refused(.truncatedFrame)")
        }
    }

    func testOverMaxLengthPrefixIsFrameTooLarge() {
        guard case .refused(.frameTooLarge) = readAfter([0xff, 0xff, 0xff, 0xff]) else {
            return XCTFail("expected .refused(.frameTooLarge)")
        }
    }
}

File: protocol-generator/csharp/test/EntityCore.Protocol.Tests/ScopeAlgebraTests.cs
using EntityCore.Protocol.Capability;
using EntityCore.Protocol.Codec;
using EntityCore.Protocol.Model;
using Xunit;

namespace EntityCore.Protocol.Tests;

/// <summary>
/// The §5 scope algebra at the unit level: §5.2's effective-target projection (0.8.2.20,
/// N11), §6.3's <c>check_path_permission</c>, the §5.4 sentinel's scoping to PATH-SCOPE
/// (0.8.2.24 N2/N3) and §5.5a's <c>scope_subset</c> typed by scope kind (F50, ruled at
/// 0.8.2.16).
/// <para>
/// AUTHORED HERE BECAUSE THE PINNED CHECK SET HAS NO VECTOR ON THIS SURFACE. Every one of
/// these behaviours is measurable on the wire by <c>tools/arc-probe</c> and by nothing in
/// the 778-check core gate, so a unit that regresses silently is exactly what
/// <c>run-s2.sh</c> is for.
/// </para>
/// </summary>
public sealed class ScopeAlgebraTests
{
    private const string Local = "z6MkLocalPeerIdentityBase58PlaceholderAAAAAAAAAAA";

    // ── §5.2 effective targets (0.8.2.20 / N11) ────────────────────────────────────

    /// <summary>
    /// The pair — survivors AND "was there a resource at all" — is §3.3's NON-LOSSY
    /// PROJECTION <c>[MUST]</c> (0.8.2.25, N11). A function returning only a list collapses
    /// <c>[qA] exclude [qA]</c> to <c>[]</c> and deletes the two-empties discriminator
    /// before any handler can read it.
    /// </summary>
    [Fact]
    public void EffectiveTargets_SeparatesTheTwoEmpties()
    {
        Permissions.EffectiveTargetSet absent = Permissions.EffectiveTargets(ExecuteWith(null), Local);
        Assert.False(absent.HasResource);
        Assert.Empty(absent.Survivors);

        Permissions.EffectiveTargetSet selfExcluded = Permissions.EffectiveTargets(
            ExecuteWith(Resource(new[] { "qA" }, new[] { "qA" })), Local);
        Assert.True(selfExcluded.HasResource);           // the discriminator N11 requires
        Assert.Empty(selfExcluded.Survivors);
    }

    /// <summary>
    /// The caller's own exclude removes entries BEFORE anything else looks at the request,
    /// and survivors come back in the caller's OWN SPELLING — 0.8.2.21 is explicit that
    /// <c>effective_targets</c> yields raw survivors, and the value flows on to a tree
    /// lookup that canonicalizes for itself.
    /// </summary>
    [Fact]
    public void EffectiveTargets_NarrowsAndKeepsTheCallersSpelling()
    {
        Permissions.EffectiveTargetSet one = Permissions.EffectiveTargets(
            ExecuteWith(Resource(new[] { "qA", "qB" }, new[] { "qA" })), Local);
        Assert.True(one.HasResource);
        Assert.Equal(new[] { "qB" }, one.Survivors);     // raw, not "/{local}/qB"

        Permissions.EffectiveTargetSet two = Permissions.EffectiveTargets(
            ExecuteWith(Resource(new[] { "qA", "qB" }, null)), Local);
        Assert.Equal(new[] { "qA", "qB" }, two.Survivors);
    }

    /// <summary>
    /// The caller-exclude arm is fail-OPEN on an unmatchable pattern; §5.4 rules it
    /// separately from the grant arm, where the same value is fail-CLOSED. INHERITED from
    /// the primitives rather than restated: the pattern canonicalizes to the sentinel and
    /// <see cref="Paths.MatchesPattern"/> then answers false, so the target survives.
    /// </summary>
    [Fact]
    public void EffectiveTargets_UnmatchableCallerExcludeCarvesOutNothing()
    {
        Permissions.EffectiveTargetSet r = Permissions.EffectiveTargets(
            ExecuteWith(Resource(new[] { "qA" }, new[] { "../nope" })), Local);
        Assert.Equal(new[] { "qA" }, r.Survivors);
    }

    /// <summary>
    /// A PRESENT-BUT-ILL-TYPED <c>targets</c> is PRESENT, with an empty survivor list.
    /// Reporting it absent would serve the WIDER absent-case answer to a request that named
    /// a resource — N11's own defect one field over, and the divergence the two vanguards
    /// found between themselves.
    /// </summary>
    [Fact]
    public void EffectiveTargets_IllTypedTargetsAreStillPresent()
    {
        EcfValue resource = Ecf.Map(("targets", Ecf.Text("not-an-array")));
        Permissions.EffectiveTargetSet r = Permissions.EffectiveTargets(ExecuteWith(resource), Local);
        Assert.True(r.HasResource);
        Assert.Empty(r.Survivors);

        // A resource map with no `targets` key at all is ABSENT, not present-and-empty:
        // there is no target list to have been emptied.
        Permissions.EffectiveTargetSet noKey =
            Permissions.EffectiveTargets(ExecuteWith(Ecf.Map(("exclude", Ecf.Array(Ecf.Text("x"))))), Local);
        Assert.False(noKey.HasResource);
    }

    // ── §6.3 check_path_permission ─────────────────────────────────────────────────

    /// <summary>
    /// ONE ACCEPT AND ONE DENY PER DIMENSION. The accept case is what validates the
    /// FIXTURE: a predicate test built only from deny cases is indistinguishable from one
    /// asserting <c>False == False</c>, because a broken fixture denies everything for
    /// free. One deny per dimension is what says the predicate checks the dimension rather
    /// than merely being able to say no.
    /// </summary>
    [Fact]
    public void CheckPathPermission_AcceptsInGrantAndDeniesPerDimension()
    {
        CapabilityToken cap = Token(Grant(
            handlers: Scope0("system/tree"),
            resources: Scope0("q/*"),
            operations: Scope0("get")));

        Assert.True(Permissions.CheckPathPermission("get", "q/a", cap, "system/tree", Local));

        // resources: a path outside the grant.
        Assert.False(Permissions.CheckPathPermission("get", "other/a", cap, "system/tree", Local));
        // operations: id-scope, literal.
        Assert.False(Permissions.CheckPathPermission("put", "q/a", cap, "system/tree", Local));
        // handlers: a different owning handler.
        Assert.False(Permissions.CheckPathPermission("get", "q/a", cap, "system/capability", Local));
    }

    /// <summary>
    /// An empty <c>resources.include</c> is a legal grant shape (§5.2: handlers that touch
    /// no tree paths) and DENIES every path — <c>covered</c> over an empty include list is
    /// false, which is what that note says it should be.
    /// </summary>
    [Fact]
    public void CheckPathPermission_EmptyResourceIncludeDeniesEveryPath()
    {
        CapabilityToken cap = Token(Grant(
            handlers: Scope0("system/tree"),
            resources: new Scope(System.Array.Empty<string>(), null),
            operations: Scope0("get")));
        Assert.False(Permissions.CheckPathPermission("get", "q/a", cap, "system/tree", Local));
    }

    /// <summary>
    /// Canonicalization is TOTAL in a matcher position (0.8.2.20): a malformed path answers
    /// the sentinel, which matches no grant, so it falls through to DENY rather than
    /// escaping as a throw the §6.5 frame would answer with the ADMISSION disposition.
    /// </summary>
    [Fact]
    public void CheckPathPermission_MalformedPathDeniesRatherThanThrowing()
    {
        CapabilityToken cap = Token(Grant(
            handlers: Scope0("system/tree"),
            resources: Scope0("*"),
            operations: Scope0("get")));
        // The wide-open control first, or "false everywhere" would satisfy this vacuously.
        Assert.True(Permissions.CheckPathPermission("get", "q/a", cap, "system/tree", Local));
        Assert.False(Permissions.CheckPathPermission("get", "q//a", cap, "system/tree", Local));
        Assert.False(Permissions.CheckPathPermission("get", "../escape", cap, "system/tree", Local));
    }

    // ── §5.4 sentinel, SCOPED TO PATH-SCOPE (0.8.2.24 N2/N3) ───────────────────────

    /// <summary>
    /// <em>"a capability carrying an unmatchable PATH-SCOPE pattern is INVALID … It does
    /// NOT reach <c>operations</c> or <c>peers</c> <c>[MUST]</c>"</em>.
    /// <para>
    /// The unscoped guard ran an id pattern through the §5.4 transforms purely to classify
    /// it and then denied the WHOLE dimension on a property unrelated to whether the
    /// exclude carves anything out: <c>*/apply</c>, an ordinary namespaced operation name,
    /// path-canonicalizes to the sentinel. Over-denial, invisible on any well-formed grant.
    /// </para>
    /// </summary>
    [Fact]
    public void Sentinel_DeniesOnPathScopeOnly()
    {
        var idScope = new Scope(new[] { "*" }, new[] { "*/apply" });
        // The ID arm: `*/apply` is a literal that simply does not equal `compute`, so the
        // include still carries and the dimension is NOT denied.
        Assert.True(idScope.Matches("compute", Local, ScopeKind.Id));
        // …and it still EXCLUDES what it literally names, or the guard has merely been
        // deleted rather than scoped.
        Assert.False(idScope.Matches("*/apply", Local, ScopeKind.Id));

        // The PATH arm keeps 0.8.2.21's rule: an unmatchable exclude DENIES rather than
        // carving out nothing.
        var pathScope = new Scope(new[] { "*" }, new[] { "../nope" });
        Assert.False(pathScope.Matches("q/a", Local, ScopeKind.Path));
        // Control: the identical include with no unmatchable exclude allows.
        Assert.True(new Scope(new[] { "*" }, null).Matches("q/a", Local, ScopeKind.Path));
    }

    // ── §5.5a scope_subset typed by scope kind (F50 / 0.8.2.16) ────────────────────

    /// <summary>
    /// §3.6's id-scope grammar binds the scope TYPE, not one function — <em>"An
    /// implementation on the canonicalizing reading is non-conformant and MUST adopt the
    /// literal matcher"</em> — so the rule F40 landed on <see cref="Scope.Matches"/> reaches
    /// <c>scope_subset</c> too.
    /// <para>
    /// <c>*/apply</c> is the witness and it is fail-CLOSED: under the canonicalizing
    /// reading it becomes the §5.4 sentinel, which matches nothing in either operand, so a
    /// legitimate child <c>operations</c> include was refused against a parent of <c>*</c>.
    /// The literal matcher covers it.
    /// </para>
    /// </summary>
    [Fact]
    public void ScopeSubset_OperationsUseTheLiteralMatcher()
    {
        GrantEntry wideOpen = Grant(Scope0("*"), Scope0("*"), Scope0("*"));

        GrantEntry namespacedOp = Grant(Scope0("system/tree"), Scope0("q/*"), Scope0("*/apply"));
        Assert.True(Attenuation.GrantsWithinAuthority(new[] { namespacedOp }, new[] { wideOpen }, Local));

        // The EXCLUDE arm of the same pair: a parent exclude the child repeats verbatim is
        // inherited under the literal matcher and was not under the canonicalizing one.
        GrantEntry parentEx = Grant(Scope0("*"), Scope0("*"),
            new Scope(new[] { "*" }, new[] { "*/apply" }));
        GrantEntry childEx = Grant(Scope0("system/tree"), Scope0("q/*"),
            new Scope(new[] { "*" }, new[] { "*/apply" }));
        Assert.True(Attenuation.GrantsWithinAuthority(new[] { childEx }, new[] { parentEx }, Local));

        // THE MATCHER HALF, and it is here because its plant RAN GREEN without it. The
        // `*/apply` cases above are decided by the FRAME (the sentinel), so replacing only
        // the `Covers` local with the path matcher left every assertion passing — an inert
        // control. This pair separates the two matchers on strings the frame leaves alone:
        // §5.4's peer-wildcard walk makes `MatchesPattern("/x/get", "/*/get")` TRUE while
        // §3.6's literal matcher answers FALSE. Under the canonicalizing reading the child
        // therefore reads as covered by a parent it does not literally match, and the child
        // grant comes out WIDER than its parent — the delegation-chain widening F50 names.
        GrantEntry peerWildcardOp = Grant(Scope0("system/tree"), Scope0("q/*"), Scope0("/x/get"));
        GrantEntry peerWildcardParent = Grant(Scope0("*"), Scope0("*"), Scope0("/*/get"));
        Assert.False(
            Attenuation.GrantsWithinAuthority(new[] { peerWildcardOp }, new[] { peerWildcardParent }, Local),
            "operations must use the LITERAL matcher, not the peer-wildcard walk");

        // CONTROL — the typing must not turn scope_subset into a rubber stamp: an
        // operations include the parent does not literally cover is still refused.
        GrantEntry notCovered = Grant(Scope0("system/tree"), Scope0("q/*"), Scope0("put"));
        GrantEntry narrowParent = Grant(Scope0("*"), Scope0("*"), Scope0("get"));
        Assert.False(Attenuation.GrantsWithinAuthority(new[] { notCovered }, new[] { narrowParent }, Local));

        // CONTROL — the PATH dimensions keep the canonicalizing matcher: a resources
        // include outside the parent's subtree is refused.
        GrantEntry widerResource = Grant(Scope0("system/tree"), Scope0("other/*"), Scope0("get"));
        GrantEntry narrowResource = Grant(Scope0("*"), Scope0("q/*"), Scope0("*"));
        Assert.False(Attenuation.GrantsWithinAuthority(new[] { widerResource }, new[] { narrowResource }, Local));
    }

    // ── fixtures ───────────────────────────────────────────────────────────────────

    private static Scope Scope0(params string[] include) => new(include, null);

    private static GrantEntry Grant(Scope handlers, Scope resources, Scope operations) =>
        new(handlers, resources, operations, null, null, null);

    private static CapabilityToken Token(params GrantEntry[] grants)
    {
        byte[] hash = new byte[33];
        Entity entity = Entity.Create(TypeNames.CapabilityToken, Ecf.Map(
            ("grants", Ecf.Array(grants.Select(g => g.ToEcf()))),
            ("granter", Ecf.Bytes(hash)),
            ("grantee", Ecf.Bytes(hash)),
            ("created_at", Ecf.Uint(1_700_000_000_000))));
        return new CapabilityToken(entity);
    }

    private static EcfValue Resource(IEnumerable<string> targets, IEnumerable<string>? exclude) => Ecf.Map(
        ("targets", Ecf.Array(targets.Select(Ecf.Text))),
        ("exclude", exclude is null ? null : Ecf.Array(exclude.Select(Ecf.Text))));

    /// <summary>
    /// An EXECUTE carrying <paramref name="resource"/> VERBATIM — built through
    /// <see cref="Ecf"/> rather than <see cref="ResourceTarget"/>, because that type is an
    /// ADMISSION parser and would refuse exactly the shapes these cases are about.
    /// </summary>
    private static Execute ExecuteWith(EcfValue? resource) => new(Entity.Create(TypeNames.Execute, Ecf.Map(
        ("request_id", Ecf.Text("t")),
        ("uri", Ecf.Text("system/tree")),
        ("operation", Ecf.Text("get")),
        ("resource", resource),
        ("params", Ecf.EmptyMap))));
}
